enhancements/management-workload-partitioning.md

```diff
@@ -0,0 +1,710 @@
+   management workloads in a special way.
+5. The installer creates a machine config manifest to write a
+   configuration file for kubelet. The file should only be readable by
+   the kubelet.
+6. The installer enables the feature set for the feature flag
+   controlling management workload partitioning.
+7. The kubelet starts up and finds the configuration file enabling the
+   new feature.
+8. The kubelet advertises `cpu.workload.openshift.io/management`
+   extended resources on the node based on the number of CPUs in the
+   host.
+9. The kubelet reads static pod definitions. It replaces the `cpu`
+   requests with `cpu.workload.openshift.io/management` requests of
+   the same value and adds the
+   `cpu.workload.openshift.io/{container-name}` annotations for CRI-O
+   with the same values.
+10. Something schedules a regular pod with the
+    `workload.openshift.io/target=management` label in a namespace
+    with the `workload.openshift.io/allowed=management` label.
+11. The admission hook modifies the pod, replacing the CPU requests
+    with `cpu.workload.openshift.io/management` requests and adding
+    the `cpu.workload.openshift.io/{container-name}` annotations for
+    CRI-O.
+12. The scheduler sees the new pod and finds available
+    `cpu.workload.openshift.io/management` resources on the node. The
+    scheduler places the pod on the node.
+13. Repeat steps 10-12 until all pods are running.
+14. Single-node deployment comes up with management components
+    constrained to subset of available CPUs
+
+#### Pod mutation
+
+The kubelet and API admission hook will change pods labeled with
+`workload.openshift.io/management` so the CPU requests are replaced
+with management CPU requests and an annotation is added with the same
+value.
+
+```yaml
+requests:
+  cpu:
+    400m
+```
+
+becomes
+
+```yaml
+requests:
+  cpu.workload.openshift.io/management: 400m
+```
+
+and
+
+```yaml
+annotations:
+  io.openshift.workload.management.cpu/container-name: 400m
+```
+
+As a special case, kubelet and the API admission hook should strip any
+annotations from pods trying to set the "cpuset" resource. See the
+next section for more details.
+
+#### CRI-O Changes
+
+*NOTE: This section needs to be updated to deal with the generalized
+CPU pools.*
+
+CRI-O will be updated to support new configuration settings for workload types.
+
+```ini
+[crio.runtime.workloads.{workload-type}]
+  label = "workload.openshift.io/target={workload-type}"
+  resource_annotation_prefix = "io.openshift.workload.{workload-type}"
+  resources = {
+    "cpu": "",
+    "cpuset": "0-1",
+  }
+```
+
+The `label` field is used to match pods that should be treated as
+having the workload type. The label key on the pod is compared for an
+exact match against the value specified in the configuration file.
+
+The `resource_annotation_prefix` is the start of the annotation key
+used to pass settings down to CRI-O.
+
+The `resources` map associates annotation suffixes with default
```

Comment:
Can you please clarify the difference between `cpu` and `cpuset`?
IIUC we should specify the cpuset that we will pin the relevant group of pods to it. But how will it work with the CPU shares?

Reply:
We are using the CPU requests specified in the pod manifest for the containers and converting those to CPU shares. See the kubelet PoC PR.